oauth2-store/src/main/java/org/xwiki/contrib/oidc/internal/AbstractNimbusOAuth2TokenStore.java

```diff
@@ -122,13 +159,33 @@ public OAuth2Token getToken(OIDCClientConfiguration configuration) throws OAuth2
         return getToken(getConfiguredDocumentReference(configuration), configuration);
     }
 
-    protected NimbusOAuth2Token getToken(DocumentReference documentReference,
+    protected OAuth2Token getToken(DocumentReference documentReference,
         OIDCClientConfiguration configuration)
         throws OAuth2Exception
     {
+
+        String cacheKey = serializer.serialize(documentReference);
+        Map<String, OAuth2Token> tokenCacheMap = cache.get(cacheKey);
+        String cacheMapKey = configuration.getConfigurationName();
+        if (tokenCacheMap != null && tokenCacheMap.containsKey(cacheMapKey)) {
+            return tokenCacheMap.get(cacheMapKey);
+        }
+        return getTokenFromDocument(documentReference, configuration);
+    }
+
+    private synchronized OAuth2Token getTokenFromDocument(DocumentReference documentReference,
```

Comment:
It might not be relevant here, but in case you're interested in having a bit more parallelism here (and just don't want several requests for the same document/configuration), you could use a [Striped](https://github.com/google/guava/wiki/StripedExplained) lock (Guava is already included as a dependency in XWiki, we're just not using it much).

Reply:
I'm wondering if this could become a bottleneck and if it wouldn't be okay to just remove the locking here now that you don't rely on the token instances being unique anymore. The problem with the locking here is that if a token is stored in a document that takes long to load, all other non-cached tokens cannot be loaded anymore while this one document is loaded.